Task1

mccreed13 opened the pull request.

mccreed13 committed:
Relocate files

mccreed13 committed:
Rename package

mccreed13 committed:
Moved test class to src/test/java.
Implemented page object model.
Refactored methods.
Added assertions.

vaddost commented on src/main/java/com/epam/training/student_anton_lapushenko/WebDriver/Task1/CreateNewPastePage.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    private void setText(String text) {
+        List<WebElement> textarea = driver.findElements(textAreaBy);
```
use findElement instead of findElements

vaddost commented on src/main/java/com/epam/training/student_anton_lapushenko/WebDriver/Task1/CreateNewPastePage.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    private void setExpirationTime(List<WebElement> selects, ExpirationTime expirationTime) {
+        selects.get(2).click();
```
what if the order of fields will be changed? or new field with type select will be added? 
try to find unique locators by id, xpath or css instead.
And make sure that your locator will return exactly one element

mccreed13 replied:
Is code lower, find options in this select with Enum, correct?

vaddost replied:
yes, it's correct

vaddost commented on src/main/java/com/epam/training/student_anton_lapushenko/WebDriver/Task1/CreateNewPastePage.java:
```diff
@@ -0,0 +1,56 @@
+        }
+    }
+
+    public PastePage createPaste(String text, String nameOrTitle, ExpirationTime expirationTime) {
```
this method can be moved to Business Object class with page steps. But it's ok for now

vaddost commented on src/main/java/com/epam/training/student_anton_lapushenko/WebDriver/Task1/CreateNewPastePage.java:
```diff
@@ -0,0 +1,56 @@
+
+    public CreateNewPastePage(WebDriver driver) {
+        this.driver = driver;
+        if (!driver.getCurrentUrl().equals("https://pastebin.com/")) {
```
it's not good practice to do such verifications in the constructor. In the future you can add additional assertion in your test to check the browser current location

vaddost commented on src/main/java/com/epam/training/student_anton_lapushenko/WebDriver/Task1/PastePage.java:
```diff
@@ -0,0 +1,37 @@
+        By wrapperPasteBy = By.xpath("//*[contains(@class, 'post-view')]");
+        if (driver.findElements(wrapperPasteBy).isEmpty()) {
+            throw new IllegalStateException("This is not Paste Page," +
+                    " current page is: " + driver.getCurrentUrl());
```
the same comment as for CreateNewPastePage constructor.

vaddost commented on src/test/java/com/epam/training/student_anton_lapushenko/WebDriver/Task1/BaseChromeTest.java:
```diff
@@ -0,0 +1,12 @@
+import org.junit.jupiter.api.BeforeEach;
+import org.openqa.selenium.chrome.ChromeDriver;
+
+class BaseChromeTest extends BaseTest {
```
Too much BaseTests, in my opinion. You can control the browser by some configs property or by BrowserManager that you can implement.
But Ok for this homework

mccreed13 replied:
I created 2 bases, one for all tests, one for chrome browser. I tried realise abstraction from OOP. I think this structure is good, because in future I would create many bases for many browsers/drivers. Is that good practice or not?

vaddost replied:
That's not a good approach. Let's imagine you have 4000 tests that you need to run in parallel on 3 browsers. With your approach you need to create separate tests for all browsers. And you will have a lot of duplicated tests which does not follow OOD principles.
There is a pattern, WebDriver Factory, that is generally used for this purpose. You can find one of the implementation here: https://dev.to/eliasnogueira/the-best-way-to-create-browser-instances-using-the-factory-pattern-java-and-selenium-webdriver-16l5

But you can create your own implementation, but the general idea is as described in doc.

We can have different base for different test types or for tests that need different preconditions.
But it's not a good practice to use different Base for different browser types.

vaddost reviewed (commented):
Regarding the Select, you're right, didn't metion that select field is hidden. Skip this comment
Generally acceptable for homework 1
For homework 2 please resolve my comments
You can create PR with changes as well
This one can be merged

vaddost commented:
one more comment about page structure in main
please use different packages for different entities.
So for the next task update package structure
e.g. util - for utilities classes, pages - for po, enums - for enums etc.

mccreed13 committed:
Resolved problems
In CreateNewPastePage.setText changed findElements to findElement.
Changed finds select object of expiration time.
createPaste method moved to Business Object.
Removed checks in constructors.
Realised BrowserFactory.

mccreed13 committed:
Relocate files to packages.
